### ISSUE TITLE
[BUG] Division by zero exception for vision metrics

### ISSUE THREAD

**Issue by AkshataNalavadi** [deepchecks/deepchecks#2299]
![Error_MeanAverageRecallReport](https://user-images.githubusercontent.com/83868777/213799424-34403e53-afa2-4f70-aa2d-30167258193a.png)

**Describe the bug**
I am using Vision Based Model DeepChecks on my 'Yolov5m' model using Pytorch library.
While some of the checks such as 'ClassPerformance',  'ImageSegmentPerformance', 'ConfusionMatrixReport', 'TrainTestPredictionDrift' are working and producing results, I am getting 'divide by zero' for 'MeanAverageRecallReport' and 'MeanAveragePrecisionReport'- As you can see in the error screenshot, it is giving error as num_classes is zero. But I am able to print the classes and also the check 'ClassPerformance' is already working. So I think num_classes should not be zero and I am wondering where can be the issue for these checks ?

**To Reproduce**
1. Define datasets
2. Implement Detection Data Class
3. Import backbone model
4. Transfer weights from trained model(yolov5)
5. Run Model Deepchecks
6. Referred these 2 links for my work : https://docs.deepchecks.com/stable/user-guide/vision/auto_quickstarts/plot_detection_tutorial.html
https://docs.deepchecks.com/stable/user-guide/vision/data-classes/VisionData.html

**Expected behavior**
Expecting the MeanAverageRecallReport and MeanAveragePrecisionReport to produce results

**Screenshots**
Screenshot attached

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04
 - Python Version: 3.9
 - Deepchecks Version: 0.10.0

**Comment by noamzbr** [comment]
Thanks for reporting @AkshataNalavadi! It sure looks like a bug, as your labels seem fine (and considering that ClassPerformance worked correctly for your data).

To help us debug it, can you perhaps provide a minimal reproducible example for this bug?

**Comment by AkshataNalavadi** [comment]
Hi, here is the example code I have used : 
`from deepchecks.vision import DetectionData
import torch
import numpy as np
import os
from PIL import Image
from torch.utils.data import DataLoader, Dataset
from ruamel import yaml
import random
import json



class TLDataset(Dataset):
    def __init__(self, root, yaml_location, yaml_file, data_dir, transforms,img_list):
        self.root = root
        self.transforms = transforms
        self.annotationpath = os.path.join(root, yaml_location, yaml_file)
        self.images = img_list
        self.data_dir = data_dir
        
    def __getitem__(self, idx):
        img_path = self.images[idx]
        img_name = img_path.split("/")[-1]
        print('Image Name :', img_name)
        img = Image.open(img_path).convert("RGB")
        annotations = yaml.safe_load(open(self.annotationpath))
        # Find annotations for specific image
        annotation = [element for element in annotations if element['path'] == (self.data_dir + img_name)][0]

        bboxes = []
        labels = []
        for obj in annotation['objects']:
            #Using Format [x_min, y_min, w, h, coTLDatasetnfidence, class_id]
            ##xmax = width+xmin, #ymax = height+ymin
            b = [float(obj['x']), float(obj['y']), float(obj['width']+float(obj['x'])),
                    float(obj['height']+float(obj['y']))]
            bboxes.append(b)
            class_id = str(obj['class_id'])[4]
            labels.append(int(class_id))

        print('labels:',labels)
        bboxes = torch.as_tensor(np.array(bboxes), dtype=torch.float32, device="cuda" if torch.cuda.is_available() else "cpu")
        labels = torch.as_tensor(np.array(labels), dtype=torch.int64, device="cuda" if torch.cuda.is_available() else "cpu")
        print('labels_tensors:',labels)

        if self.transforms is not None:
            res = self.transforms(image=np.array(img), bboxes=bboxes, class_labels=labels)
        else:
            res = {'image':torch.as_tensor(np.array(img), device="cpu"), 'bboxes':bboxes, 'class_labels':labels}
            # res = {'image':torch.tensor(img), 'bboxes':bboxes, 'class_labels':labels}

        target = {
            'boxes': [torch.Tensor(x) for x in res['bboxes']],
            'labels': res['class_labels']
        }

        img = res['image']

        return img, target
    
    def __len__(self):
        return len(self.images)

    

def read_label_map(label_map_path):

    item_id = None
    item_name = None
    items = {}
    
    with open(label_map_path, "r") as file:
        for line in file:
            line.replace(" ", "")
            if line == "item{":
                pass
            elif line == "}":
                pass
            elif "id" in line:
                item_id = int(line.split(":", 1)[1].strip())
            elif "name" in line:
                item_name = line.split(":", 1)[1].replace("'", "").strip()

            if item_id is not None and item_name is not None:
                items[item_name] = item_id
                item_id = None
                item_name = None

    return items

       

class ObjectDetectionClassTest(DetectionData):

    def batch_to_images(self, batch):
        """Convert a batch of images to a list of PIL images.

        Parameters
        ----------
        batch : torch.Tensor
            The batch of images to convert.

        Returns
        -------
        list
            A list of PIL images.
        """

        # Assuming batch[0] is a batch of (N, C, H, W) images, we convert it to (N, H, W, C)/
        imgs = torch.stack(list(batch[0])).cpu().detach().numpy().transpose((0,2,1,3))

        # The images are normalized to [0, 1] range based on the mean and std of the ImageNet dataset, so we need to
        # convert them back to [0, 255] range.
        mean = [0.485, 0.456, 0.406]
        std = [0.229, 0.224, 0.225]
        imgs = std * imgs + mean
        imgs = np.clip(imgs, 0, 1)
        imgs *= 255
        return imgs

    def batch_to_labels(self, batch):
        """Convert a batch bounding boxes to the required format.

        Parameters
        ----------
        batch : tuple
            The batch of data, containing images and bounding boxes.

        Returns
        -------
        List
            A list of size N containing tensors of shape (B,5).
        """

        # each bbox in the labels is (class_id, x, y, x, y). convert to (class_id, x, y, w, h)
        ##xmax-xmin = width, #ymax-ymin = height
        tensor_annotations = batch[1]
        label = []
        for annotation in tensor_annotations:
            if len(annotation["boxes"]):
                bbox = torch.stack(annotation["boxes"])
                bbox[:, 2:] = bbox[:, 2:] - bbox[:, :2]               
                # The label shape is [class_id, x, y, w, h]
                label.append(
                    torch.concat([annotation["labels"].reshape((-1, 1)), bbox], dim=1)
                )
            else:
                # If it's an empty image, we need to add an empty label
                label.append(torch.tensor([]))
        
        return label

    def infer_on_batch(self, batch, model, device):
        """Get the predictions of the model on a batch of images.

        Parameters
        ----------
        batch : tuple
            The batch of data, containing images and bounding boxes.
        model : torch.nn.Module
            The model to use for inference.
        device : torch.device
            The device to use for inference.

        Returns
        -------
        List
            A list of size N containing tensors of shape (B,6).
        """

        """
        Returns the predictions for a batch of data. The expected format is a list of tensors of shape length N, where N
        is the number of samples. Each tensor element is in a shape of [B, 6], where B is the number of bboxes in the
        predictions, and each bounding box is in the structure of [x, y, w, h, score, class_id].
        """
        return_list = []
        # predictions = model.to(device)(batch[0])
        predictions = model(img_list_test)

        # yolo Detections objects have List[torch.Tensor(B,6)] output where each bbox is
        #(x_min, y_min, x_max, y_max, confidence, class_id).
        for single_image_tensor in predictions.pred:
            pred_modified = torch.clone(single_image_tensor)
            pred_modified[:, 2] = pred_modified[:, 2] - pred_modified[:, 0]
            pred_modified[:, 3] = pred_modified[:, 3] - pred_modified[:, 1]
            return_list.append(pred_modified)
       
        return return_list
        

root=''
params = yaml.safe_load(open(root+'params.yaml'))["model_checks"]
dataset_sample_size = params['dataset_sample_size_mdl']
checks_output_dir = params['model_checks_output_dir']


yaml_file_test='data/test.yml'
annotations_test = yaml.safe_load(open(yaml_file_test))
img_list_test = random.sample([os.path.join(root, element['path']) for element in annotations_test], dataset_sample_size)



def main():


    test_set = TLDataset(root='', yaml_location = 'data',
                        yaml_file= 'test.yml', data_dir='data/images/test/',
                        transforms= None, img_list = img_list_test)
    
    test_loader = DataLoader(test_set, batch_size=2, collate_fn=(lambda batch: tuple(zip(*batch))))
    
    
    
    # Now, in order to test the class, we can create an instance of it:
   
    print('Defining the class for Object Detection:')
    test_data = ObjectDetectionClassTest(data_loader=test_loader)
    
    ##REQÙIRED MODEL
    model = torch.hub.load('ultralytics/yolov5', 'custom' , path='models/best.pt', force_reload=True)
    device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
    
    model.eval() 
    
    
    # 3. MeanAverageRecallReport
    from deepchecks.vision.checks import MeanAverageRecallReport
    check = MeanAverageRecallReport()
    result = check.run(test_data, model)
    result_MeanAverageRecallReport = result.save_as_html('MeanAverageRecallReport.html')#give proper name for html file -- DIVIDE BY ZERO
    serialized_output = result_MeanAverageRecallReport.to_json()
    result_MeanAverageRecallReport_json = json.loads(serialized_output)
    meanavgrRR_flname  = checks_output_dir + '/MeanAverageRecallReport.json'
    with open(meanavgrRR_flname, 'w') as outfile:
        json.dump(result_MeanAverageRecallReport_json, outfile)
    
    
    # 4. MeanAveragePrecisionReport
    from deepchecks.vision.checks import MeanAveragePrecisionReport
    check = MeanAveragePrecisionReport()
    result = check.run(test_data, model)
    result_MeanAveragePrecisionReport = result.save_as_html('MeanAveragePrecisionReport.html')#give proper name for html file -- DIVIDE BY ZERO
    serialized_output = result_MeanAveragePrecisionReport.to_json()
    result_MeanAveragePrecisionReport_json = json.loads(serialized_output)
    meanavgrpreR_flname  = checks_output_dir + '/MeanAveragePrecisionReport.json'
    with open(meanavgrpreR_flname, 'w') as outfile:
        json.dump(result_MeanAveragePrecisionReport_json, outfile)
    
    
if __name__ == '__main__':
    main()`

**Comment by noamzbr** [comment]
Thanks @AkshataNalavadi, is there an open dataset that we can use to run this code?

**Comment by AkshataNalavadi** [comment]
@noamzbr You can use BSTLD dataset.

**Comment by noamzbr** [comment]
Hi @AkshataNalavadi , sorry but I tried to run your code on the BSTLD dataset and run into some discrepancies between the dataset structure and the code, and couldn't get it to run.

We'd be glad to try to debug this further if you could reproduce this problem on a simpler, self-contained code sample.

**Comment by AkshataNalavadi** [comment]
Hi @noamzbr, then can you run this code using DTLD dataset. And let me know how it works. Thank you.

**Comment by AkshataNalavadi** [comment]
Hi @noamzbr, Did you try on DTLD dataset???

**Comment by matanper** [comment]
Hey @AkshataNalavadi,
we have released a new version of the vision package with drastic changes. Can you please try to convert your code (see [docs](https://docs.deepchecks.com/stable/user-guide/vision/auto_tutorials/plot_detection_tutorial.html)) and then send us the result of `ObjectDetectionClassTest(data_loader=test_loader).head()`?

If you need any help with the new structure feel free to reach us in the slack community

**Comment by noamzbr** [comment]
Resolved in new version of the vision package.

**Comment by AkshataNalavadi** [comment]
Thank you for the update @noamzbr & @matanper and I will use the latest update and continue to work on this.
